File: 송하연/9월4주차/네트워크.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        List<List<Integer>> graph = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();
        
        for(int i=0; i< n; i++){
            graph.add(new ArrayList<>());
        }
        
        for(int i=0; i< n; i++){
           for(int j=0; j<n; j++){
               if(computers[i][j]==1 && i!=j){
                   int start = i;
                   int end = j;
                   graph.get(i).add(j);
                   
               }
           }
        }
        
        //System.out.println(graph);
    
        boolean[] visited = new boolean[n]; 
        
        for(int i=0; i< n; i++){
            if(!visited[i]){
                int start = stack.push(i);
                visited[start]=true;
                answer++;
            }   
            
            while(!stack.isEmpty()){
                int now = stack.pop();
                for(int node: graph.get(now)){
                     if(!visited[node]){
                        stack.push(node);
                        visited[node]=true;
                    } 
                }
            }                       
        }
        return answer;
    }
}

File: 한승우/9월 4주차/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> graph[200];
int visited[200];

void bfs(int start){
    queue<int> q;
    
    visited[start] = 1;
    q.push(start);
    
    while(!q.empty()){
        int v = q.front();
        q.pop();
        
        for(int n: graph[v]){
            if(!visited[n]){
                visited[n] = 1;
                q.push(n);
            }
        }
    }
}

int solution(int n, vector<vector<int>> computers) {    
    int answer = 0;
    
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            if(computers[i][j]){
                graph[i].push_back(j);
                graph[j].push_back(i);
            }
    
    for(int i = 0; i < n; i++){
        if(!visited[i]){
            bfs(i);
            answer++;
        } 
    }
    
    return answer;
}